File: observalog-go/middleware/http.go
package middleware

import (
	"net/http"
	"strings"
	"sync"
	"sync/atomic"

	"github.com/google/uuid"
	log "github.com/darshanredkar11/observalog/observalog-go"
)

var (
	routeStages = make(map[string]string)
	mu           sync.RWMutex
)

// Middleware wraps an HTTP handler to inject trace context.
//
// Inbound propagation: if an upstream service (or API gateway) passes
// X-Trace-Id / X-Parent-Span-Id headers the trace chain is continued.
// If there are no incoming headers this is a fresh entry point and a new
// trc_<id> is generated so that ALL log calls within this request share
// the same trace_id (relying on ApplyGracefulDegradation would generate a
// different sys_<uuid> for every emitLine call).
func Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ctx := r.Context()

		// Continue an existing trace OR start a fresh one at the entry point.
		traceID := r.Header.Get("X-Trace-Id")
		if traceID == "" {
			// "trc_" (4) + 10 hex chars = 14 bytes — fits the Part A trace_id slot.
			traceID = "trc_" + uuid.New().String()[:10]
		}
		ctx = log.WithTraceID(ctx, traceID)

		// Generate a new span_id for this service hop.
		spanID := "spn_" + uuid.New().String()[:3]
		ctx = log.WithSpanID(ctx, spanID)

		// The upstream span becomes our parent_span (shows call hierarchy in the brain).
		if parentSpan := r.Header.Get("X-Parent-Span-Id"); parentSpan != "" {
			ctx = log.WithParentSpan(ctx, parentSpan)
		}

		// Propagate user identity across the chain.
		if userID := r.Header.Get("X-User-Id"); userID != "" {
			ctx = log.WithUserID(ctx, userID)
		}

		// journey_stage: prefer inherited value, fall back to auto-derive.
		journeyStage := r.Header.Get("X-Journey-Stage")
		if journeyStage == "" {
			journeyStage = deriveJourneyStage(r)
		}
		ctx = log.WithJourneyStage(ctx, journeyStage)

		// Fresh seq counter — resets to 1 for this service hop.
		ctx = log.WithSeq(ctx, new(atomic.Uint32))

		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

// deriveJourneyStage looks up the route in config or auto-derives from request path.
func deriveJourneyStage(r *http.Request) string {
	// Try config lookup first.
	mu.RLock()
	if stage, ok := routeStages[r.Method+" "+r.URL.Path]; ok {
		mu.RUnlock()
		return stage
	}
	mu.RUnlock()

	// Auto-derive from path: e.g., /api/v1/docs → api.docs
	path := strings.TrimPrefix(strings.TrimPrefix(r.URL.Path, "/api"), "/v1")
	path = strings.Trim(path, "/")
	parts := strings.Split(path, "/")
	if len(parts) == 0 || parts[0] == "" {
		return "http.root"
	}
	// Join with dots: /users/{id}/edit → http.users.edit
	return "http." + strings.Join(parts, ".")
}

// LoadRouteStages loads a route → journey_stage mapping (stub for YAML file).
// For now, this is a placeholder. In production, parse route_stages.yaml.
func LoadRouteStages(path string) error {
	// TODO: Parse YAML and populate routeStages map.
	return nil
}

File: observalog-node/src/middleware/http.ts
import { v4 as uuidv4 } from 'uuid';
import { runWithContextAsync } from '../context';

type Request  = { headers: Record<string, string | string[] | undefined>; path: string };
type Response = Record<string, unknown>;
type NextFn   = () => Promise<void> | void;

// Express-compatible middleware that injects trace context into AsyncLocalStorage.
export function observalogMiddleware(
    req: Request,
    _res: Response,
    next: NextFn,
): void | Promise<void> {
    const h = (name: string): string => {
        const v = req.headers[name.toLowerCase()];
        return Array.isArray(v) ? v[0] : (v ?? '');
    };

    // Continue an existing trace OR start a fresh one at the entry point.
    // Never leave traceId empty — applyGracefulDegradation generates a NEW
    // sys_<uuid> for every emitLine call, so all logs in one request would
    // have different trace IDs and the brain could not correlate them.
    const incomingTraceId = h('x-trace-id');
    const traceId = incomingTraceId !== ''
        ? incomingTraceId
        : 'trc_' + uuidv4().replace(/-/g, '').substring(0, 10); // 14 chars: trc_ + 10

    // Prefer inherited journey_stage; fall back to auto-derive.
    const incomingStage = h('x-journey-stage');

    const ctx = {
        traceId,
        spanId:       'spn_' + uuidv4().substring(0, 3),
        parentSpan:   h('x-parent-span-id'),  // upstream span → our parent_span
        userId:       h('x-user-id'),
        journeyStage: incomingStage !== '' ? incomingStage : deriveJourneyStage(req.path),
        seq:          { value: 0 },
    };

    const result = next();
    if (result instanceof Promise) {
        return runWithContextAsync(ctx, () => result);
    }
    runWithContextAsync(ctx, async () => {}); // sync path — seed context then return
    return result;
}

// Auto-derives journey_stage from request path.
// /api/v1/docs/123/edit → http.docs.edit (strips /api, /v1, drops numeric/UUID segments)
function deriveJourneyStage(path: string): string {
    let p = path;
    for (const strip of ['/api', '/v1']) {
        if (p.startsWith(strip)) p = p.slice(strip.length);
    }
    const parts = p.split('/').filter(s =>
        s.length > 0 && !/^([0-9a-f]{8,}|[0-9]+)$/i.test(s),
    );
    return parts.length > 0 ? `http.${parts.join('.')}` : 'http.root';
}
